lib/katello/tasks/clean_installed_packages.rake

```diff
@@ -23,17 +25,35 @@ namespace :katello do
       count += 1
     end
 
+    cleaning_total = sprintf('%.2f', Time.now - cleaning_start)
+    puts "katello_host_installed_packages table has been cleaned, total time was #{cleaning_total} seconds."
+    puts "Removing records from katello_installed_packages that are no longer referenced..."
+    cleaning_start = Time.now
+
     existing = Katello::HostInstalledPackage.select(:installed_package_id).uniq.pluck(:installed_package_id)
 
+    interrupted_run = false
     batch = 20_000
     until Katello::InstalledPackage.where('id not in (?)', existing).limit(1).count == 0
-      deleted = Katello::InstalledPackage.where('id not in (?)', existing).where("id < #{batch}").delete_all
-      sleep(3) if deleted > 0
-      puts "Batch of installed packages deleted, #{Katello::InstalledPackage.count} remaining in total"
+      begin
+        deleted = Katello::InstalledPackage.where('id not in (?)', existing).where("id < #{batch}").delete_all
+        sleep(3) if deleted > 0
+        puts "Inspected #{batch} records in katello_host_installed_packages to remove unreferenced entries in katello_installed_packages.."
+      rescue ActiveRecord::InvalidForeignKey
+        puts "Some records in this batch were unable to be removed. This is usually due to system registrations or updates that occurred after the script started."
+        puts "Cleanup will continue, but please re-run this script later (possibly during a more quiet time for the system) to fully clean set of packages."
+        interrupted_run = true
+      end
       batch += 20_000
     end
+    cleaning_total = sprintf('%.2f', Time.now - cleaning_start)
+    puts "katello_installed_packages table has been cleaned to remove unreferenced entries, total time was #{cleaning_total} seconds."
 
-    Setting[:bulk_query_installed_packages] = true
+    unless interrupted_run
+      puts "Script is complete, setting installed package search to updated method. To disable this, set 'bulk query installed packages' to 'false' in Katello settings."
+      Setting[:bulk_query_installed_packages] = true
+    end
```

Comment:
I would also re-emphasize that they need to re-run the script if interrupted_run is true

Reply:
otherwise looks good!